### ISSUE TITLE
Configurable symlink syncing

### ISSUE THREAD

**Issue by iranzo** [syncthing/syncthing#92]
According to documentation:
- Synchronizing multiple unrelated directory trees by following symlinks directly below the repository level.

This should be a bug instead of a feature, as it breaks, for example, references to other folders for completeness and instead walks trough them.

I've seen this discussion with other tools ande more or less all of them just copy the symlink instead of following it, and instead, suggest to use different folder pairs for that.

Could a configurable option be put in place to "FollowSymlinks" or to not follow them and just copy the symlink instead?

Thanks

**Comment by nille02** [comment]
Does the Solution in Go 1.4 need admin privileges? Today i see the first time a Meassage from Syncthing about the Symlinks and that he maybe need admin privileges for this oO. Even with Admin privileges the symlinks are ignored :/

But like calmh says. You can not satisfy everyone. But there is a need for a Solution that fits for the most people. 
In my Point of view. I what all files Synchronized, in this path that i provide Syncthing. For me its doesn't matter if the folder is a symlink or real folder, as long as the content is synced with each shares. If its symlink to a file is not necessary, as long the content of the file is everywhere the same.  

Like i see in this Thread, thats not fit for everyone. At the end its a matter for that purpose you use syncthing. As a Backup Software or a Sync Software.

**Comment by AudriusButkevicius** [comment]
Creating symlinks in Windows in general regardless if you use syncthing or not requires either the CreateSymlinks privilege for the given user, or administrator privileges if you don't want to go editing privileges for individual users. Hence why we require it.

When you have the privilege, links should not be ignored, and you should not get warnings.

**Comment by nille02** [comment]
Thats curious. I try to investigate is bit and open a issue.

**Comment by AudriusButkevicius** [comment]
If you do get warnings, then it means we don't have the privilege.
Also, this is not supported on old Windows which doesn't even have the feature (such as XP)

**Comment by nille02** [comment]
I have full access to the directory and its a recent windows (8.1). I can also create symlinks without any troubles. 

I get the message today (befor the .23 Update) and in the last weeks there are no problems about this.

EDIT: I find a Invalid User ID as the Owner of the Files. From the ACLs i have full access. So i added me back as the file owner but he still comes with the message. 
It would be helpful to see where he has no permissions.

**Comment by AudriusButkevicius** [comment]
We test symlink availability by trying to create one in the temp directory upon starting, this way we decide if they are supported or not.

If they are not supported, we only throw a warning when we first encounter a symlink.

**Comment by nille02** [comment]
Theres must be something broken for windows then. even in the tmp directory i can create symlinks and junctions for files and directory as i wish. and all without admin privileges.

**Comment by AudriusButkevicius** [comment]
Well then there must be something broken with your configuration.
Probably you have UAC disabled or some other non-recommended magic.

It definitely requires administrator privileges or that specific privilege on a clean Windows machine.
Are you using `mklink` to actually create it?

**Comment by nille02** [comment]
Nothing changed and i created the links with a explorer shell extension. with mklink i get a permission denied. that weired, because the folders are all owned by my user account. 

Create a junction with mklink is also not a problem, only the symlink. 

I've looked a bit about this and find a solution for create the symlinks as a user. For this you has to add the user or group to the local security polices (Control Panel -> Administrative Tools -> Local Security Policy -> Local Policies -> User Rights Assignment -> Create symbolic links) But thats not working if you are a Administrator and UAC is active ... 

A Workaround for a directory would be a junction point (my links are mostly junctions, because they point to another harddrives)

long story short. why do i get the symlink message today and not some weeks ago where i start using syncthing?

**Comment by AudriusButkevicius** [comment]
I bet the explorer shell extension has a shield icon next to it, which means elevate privileges to admin.
Junction is not a symlink.

Yeah, as I said, you can adjust individual permissions on users to go around the problem, aka allow simple user to create symlinks.

The reason you are getting messages this week but not last week is most likely because we have not stumbled across a symlink in your set of files before, hence no warnings.

**Comment by nille02** [comment]
No, its not comes with a UAC https://www.dropbox.com/s/32pkqawzg13zyz8/Screenshot%202015-02-16%2016.12.36.png?dl=0

And i know that a junction is not a symlink but for directory is comes as close as possible and has some benefits. 

There is nothing new in the Share. The Links are present since beginning and today i created only a new one.

**Comment by AudriusButkevicius** [comment]
Junction has no comparable concept on Linux, hence it's not synced.
As for why it wasn't complaining previous, there could be a million of reasons.
1. You ran an old version
2. You had symlink files ignored via .stignores
3. You were running as admin
4. You hadn't rescanned for ages
... the list goes on ...

I will not be able to tell you what you have done for it to happen.

If you don't like it, you can switch off symlinks (as well as the warning) by editing the config XML file.

**Comment by nille02** [comment]
Why should they not has a comparable concept? Both come to the same result. 

Ignoring Symlinks is also not a solution. That would exclude many Files. 

Start Syncthing as Admin would not a problem, but on the other Machines there comes a error: puller: final: creating symlink: invalid argument

**Comment by jpjp** [comment]
@nille02 any chance you could offer a patch?

**Comment by PaBLoX-CL** [comment]
For the ones that doesn't care about windows, and want to workaround about this, a nice solution is to use `mount --bind`, obviously this only works for directories, but it has allowed me to create "stacks" of different documents.

For example I create:

```
mount --bind ~/code/rails/app1 ~/code/sync/app1
mount --bind ~/code/js/another_app ~/code/sync/another_app
```

This is far from perfect and how easy is to manipulate symlinks (beginning from the fact that you need root access for bind mounts).

Then you can "sync" only `~/code/sync`. Obviously, you have to add it to fstab later, but it's quite easy and it can be easy automatized. 

    man -P 'less +/The\ bind\ mounts' mount

PS.- Also, for unique files, I believe hardlinks should work too...

**Comment by AudriusButkevicius** [comment]
This is already implemented, so workarounds not required.

**Comment by PaBLoX-CL** [comment]
I just tried to sync a symlink, and I got an error because it was a symlink O.o

**Comment by AudriusButkevicius** [comment]
What does the error say?

**Comment by PaBLoX-CL** [comment]
I don't remember exactly, I will check it on a couple of days when I get access to that machine.

**Comment by AudriusButkevicius** [comment]
If it's windows, it needs admin privileges.
